agent/pkg/auth/token_auth.go

```diff
@@ -0,0 +1,53 @@
+package auth
+
+import (
+	"context"
+	"fmt"
+
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/clientcredentials"
+	"google.golang.org/grpc/credentials"
+)
+
+// TokenPerRPCCreds satisfies the credentials.PerRPCCredentials interface
+// attaching an up to date token to each RPC request in the GRPC pipeline
+// When the token is expired a new token is requested by using the client
+// credentials flow
+type TokenPerRPCCreds struct {
+	secure bool
+	ts     oauth2.TokenSource
+}
+
+func NewTokenPerRPCCreds(clientID, clientSecret, apiAddr, agentID string, secure bool) credentials.PerRPCCredentials {
+	tokenAddrScheme := "http"
+	if secure {
+		tokenAddrScheme = "https"
+	}
+
+	authConfig := &clientcredentials.Config{
+		ClientID:     clientID,
+		ClientSecret: clientSecret,
+		TokenURL:     fmt.Sprintf("%s://%s/agent-token?agent_id=%s", tokenAddrScheme, apiAddr, agentID),
+		AuthStyle:    oauth2.AuthStyleInHeader,
+	}
+
+	return &TokenPerRPCCreds{
+		secure: secure,
+		ts:     authConfig.TokenSource(context.Background()),
+	}
+}
+
+func (t *TokenPerRPCCreds) GetRequestMetadata(ctx context.Context, uri ...string) (map[string]string, error) {
+	token, err := t.ts.Token()
+	if err != nil {
+		return nil, err
+	}
+
+	return map[string]string{
+		"authorization": token.Type() + " " + token.AccessToken,
+	}, nil
+}
```

Comment:
I am not entirely sure about this but it did not work during local developement. I think that what this does is that it ensure that we have TLS in place. I am not sure we need that. Do we need it? Basically, we can add this check there and only perform it if `secure == true`, because otherwise, it errors out as the security level is not .... secure? WDYT? In any case, this errors if we run it in local mode without TLS. This is I think why by default  `RequireTransportSecurity` of their impl always returns true...  WDYT?

Reply:
yeah, when `secure == true`, i think it makes sense to use `google.golang.org/grpc`'s implementation, then fall back to our own if `!secure`:
```diff
diff --git a/agent/pkg/auth/token_auth.go b/agent/pkg/auth/token_auth.go
index df331be..83ccb72 100644
--- a/agent/pkg/auth/token_auth.go
+++ b/agent/pkg/auth/token_auth.go
@@ -5,6 +5,7 @@ import (

        "golang.org/x/oauth2"
        "google.golang.org/grpc/credentials"
+       "google.golang.org/grpc/credentials/oauth"
 )

 // TokenPerRPCCreds satisfies the credentials.PerRPCCredentials interface
@@ -13,17 +14,21 @@ import (
 // credentials flow.
 type TokenPerRPCCreds struct {
        secure bool
-       ts     oauth2.TokenSource
+       ts     oauth.TokenSource
 }

 func NewTokenPerRPCCreds(ts oauth2.TokenSource, secure bool) credentials.PerRPCCredentials {
        return &TokenPerRPCCreds{
-               ts:     ts,
+               ts:     oauth.TokenSource{TokenSource: ts},
                secure: secure,
        }
 }

 func (t *TokenPerRPCCreds) GetRequestMetadata(ctx context.Context, uri ...string) (map[string]string, error) {
+       if t.secure {
+               return t.ts.GetRequestMetadata(ctx, uri...)
+       }
+
        token, err := t.ts.Token()
        if err != nil {
                return nil, err
```